spec/p2p/v0.34/switch.md

```diff
@@ -0,0 +1,239 @@
+error, the switch gives up reconnecting to the peer.
+
+> // ErrCurrentlyDialingOrExistingAddress indicates that we're currently
+> // dialing this address or it belongs to an existing peer.
+
+A first set of connection attempts is done at (about) regular intervals.
+More precisely, between two attempts, the switch waits for a interval of
+`reconnectInterval`, hard-coded to 5 seconds, plus a random jitter up to
+`dialRandomizerIntervalMilliseconds`, hard-coded to 3 seconds.
+At most `reconnectAttempts`, hard-coded to 20, are made using this
+regular-interval approach.
+
+A second set of connection attempts is done with exponentially increasing
+intervals.
+The base interval `reconnectBackOffBaseSeconds` is hard-coded to 3 seconds,
+which is also the increasing factor.
+The exponentially increasing dialing interval is adjusted as well by a random
+jitter up to `dialRandomizerIntervalMilliseconds`.
+At most `reconnectBackOffAttempts`, hard-coded to 10, are made using this
+exponential-intervals approach.
+
+> FIXME: the first sleep interval, to which a random jitter is applied, is 1,
+> not `reconnectBackOffBaseSeconds`, as the first exponent is `0`...
+
+## Dialing peers
+
+Dialing a peer is implemented by the `DialPeerWithAddress` method.
+It receives a peer address (`NetAddress` type), which also encodes the peer ID.
+
+The switch keeps a single dialing routine per peer ID.
+This is ensured by keeping a synchronized map `dialing` with the IDs of peers
+to which the peer is dialing.
+A peer ID is added to `dialing` when the `DialPeerWithAddress` method is called
+for that peer, and it is removed when the method returns for whatever reason.
+The method returns immediately when invoked for a peer which ID is already in
+the `dialing` structure.
+
+The actual dialing is implemented by the `Dial` method of the `Transport`
+configured for the switch, invoked in the `addOutboundPeerWithConfig` method.
+If the transport succeeds establishing a connection, the returned `Peer` is
+added to the switch using the `addPeer` method.
+This operation can fail, returning an error. In this case, the switch invokes
+the `Transport.Cleanup` method to clean any resources associated with the peer.
+
+> TODO: cleaning up a peer should mean, in practice, disconnecting from it.
+
+If the transport fails to establish a connection with the peer, and the dialed
+peer was configured as a persistent peer, the switch spawns a `reconnectToPeer`
+routine for the peer.
+If the peer is already in the `reconnecting` state, the spawned routine has no
+effect and returns immediately.
+This is in fact a likely scenario, as the `reconnectToPeer` routine relies on
+this same `DialPeerWithAddress` method for dialing peers.
+
+### `DialPeersAsync` 
+
+This method receives a list of peer addresses (strings) and dials all of
+them in parallel. 
+
+> TODO: who invokes this? Apparently, it is invoked at startup to dial all
```

Comment:
Yes these are the only places where this is called. The PEX reactor calls `DialPeerWithAddress`.

Reply:
So, we should document it as startup procedure.